lib/timber/transports/http.ex

```diff
@@ -0,0 +1,131 @@
+defmodule Timber.Transports.HTTP do
+  @moduledoc """
+  An efficient HTTP transport that buffers and delivers log messages over HTTP to the
+  Timber API. It uses batching, keep-alive connections, and msgpack to deliver logs with
+  high-throughput and little overhead.
+
+  ## Configuration
+
+  ### Custom HTTP client
+
+  By default, hackney is used via `Timber.Transports.HTTP.HackneyClient`. You can define your
+  own HTTP client by defining it in your configuration:
+
+  ```
+  config :timber, :http_transport, http_client: MyHTTPClient
+  ```
+
+  Your client *must* implement the `Timber.Transports.HTTP.Client` beahvior. Please see that
+  module for details.
+  """
+
+  @behaviour Timber.Transport
+
+  alias Timber.LogEntry
+
+  @typep t :: %__MODULE__{
+    api_key: String.t,
+    buffer_size: non_neg_integer,
+    buffer: [] | [IO.chardata],
+    flush_interval: non_neg_integer,
+    max_buffer_size: pos_integer
+  }
+
+  # 5000 log lines should be well below 1mb and provide plenty of time to handle
+  # network failures, etc.
+  @default_max_buffer_size 5000
+  @default_flush_interval 1000
+  @default_http_client __MODULE__.HackneyClient
+  @url "https://api.timber.io/frames"
+
+  defstruct api_key: nil,
+            buffer_size: 0,
+            buffer: [],
+            flush_interval: @default_flush_interval,
+            max_buffer_size: @default_max_buffer_size
+
+  @doc false
+  @spec init() :: {:ok, t}
+  def init() do
+    config =
+      config()
+      |> Keyword.put(:api_key, Timber.Config.api_key!())
```

Comment:
This depends on if we want to play nice or not. If we raise an exception here, everything goes sky high. That may be the best option since all the requests will fail.

However, it would be more in the spirit of Elixir to fail startup by returning `{:error, :no_api_key}`. The difference here is that the `Logger` application gets to decide how to handle this. Currently, `Logger` will stop when the return value of our init function isn't `{:ok, term}` ([see here](https://github.com/elixir-lang/elixir/blob/v1.4.1/lib/logger/lib/logger/watcher.ex#L16))